app/src/main/kotlin/compiler/intermediate_form/Linearization.kt

```diff
@@ -1,10 +1,111 @@
 package compiler.intermediate_form
 
 import compiler.common.intermediate_form.Covering
+import compiler.common.reference_collections.referenceHashMapOf
+import compiler.intermediate_form.Instruction.UnconditionalJumpInstruction.Jmp
 
 object Linearization {
     fun linearize(
         cfg: ControlFlowGraph,
         covering: Covering,
-    ): List<Asmable> = TODO()
+    ): List<Asmable> {
+        if (cfg.entryTreeRoot == null)
+            return emptyList()
+
+        val instructions = mutableListOf<Asmable>()
+        val labels = referenceHashMapOf<IFTNode, String>()
+        val endLabel = "end"
+        val usedLabels = mutableSetOf<String>()
+
+        fun assignLabel(node: IFTNode): String {
+            val label = "_" + labels.size
+            labels[node] = label
+            return label
+        }
+
+        fun addLabel(label: String) {
+            instructions.add(Label(label))
+        }
+
+        fun dfs(node: IFTNode, nextLabel: String) {
+            addLabel(labels[node] ?: assignLabel(node))
+
+            fun addUncoditional() {
+                instructions.addAll(covering.coverUnconditional(node))
+            }
+
+            fun addConditional(whenTrue: Boolean, targetLabel: String) {
+                instructions.addAll(covering.coverConditional(node, targetLabel, !whenTrue))
+                usedLabels.add(targetLabel)
+            }
+
+            fun addJump(targetLabel: String) {
+                if (targetLabel != nextLabel) {
+                    instructions.add(Jmp(targetLabel))
+                    usedLabels.add(targetLabel)
+                }
+            }
+
+            if (node in cfg.unconditionalLinks) {
+                val target = cfg.unconditionalLinks[node]!!
+
+                addUncoditional()
+
+                if (target in labels)
+                    addJump(labels[target]!!)
+                else
+                    dfs(target, nextLabel)
+            } else if (node in cfg.conditionalTrueLinks && node in cfg.conditionalFalseLinks) {
+                val targetWhenTrue = cfg.conditionalTrueLinks[node]!!
+                val targetWhenFalse = cfg.conditionalFalseLinks[node]!!
+
+                if (targetWhenTrue in labels) {
```

Comment:
A matter of taste, but I would divide this into 4 conditions to improve readability.